pkg/trace/agent/tags.go

```diff
@@ -251,54 +250,91 @@ func FilterTags(tags, groups []string) []string {
 // backend requirements
 // taken from dd-go.model.NormalizeTag
 func NormalizeTag(tag string) string {
-	// unless you just throw out unicode, this is already as fast as it gets
-
-	buf := bytes.NewBuffer(make([]byte, 0, 2*len(tag)))
-	lastWasUnderscore := false
-
-	for _, c := range tag {
-		// fast path for len check
-		if buf.Len() >= maxTagLength {
+	var (
+		trim   int      // start character (if trimming)
+		wiping bool     // true when the previous character has been discarded
+		wipe   [][2]int // sections to discard: (start, end) pairs
+		chars  int      // number of characters processed
+	)
+	var (
+		i int  // current byte
+		c rune // current rune
+	)
+	norm := []byte(tag)
+	for i, c = range tag {
+		if chars >= maxTagLength {
+			// we've reached the maximum
 			break
 		}
-		// fast path for ascii alphabetic chars
+		// all letters are ok
 		switch {
 		case c >= 'a' && c <= 'z':
-			buf.WriteRune(c)
-			lastWasUnderscore = false
+			chars++
+			wiping = false
 			continue
 		case c >= 'A' && c <= 'Z':
-			c -= 'A' - 'a'
-			buf.WriteRune(c)
-			lastWasUnderscore = false
+			// lower-case
+			norm[i] += 'a' - 'A'
+			chars++
+			wiping = false
 			continue
 		}
 
 		c = unicode.ToLower(c)
 		switch {
-		// handle always valid cases
 		case unicode.IsLetter(c) || c == ':':
```

Comment:
We can merge the 1st and 3rd case together?

Reply:
Actually there is a catch here: switch case statements are [evaluated top-to-bottom](https://golang.org/ref/spec#Switch_statements) and the idea here is that you can not go past the second case (length = 0) if the character isn't a letter or a colon. In other words only letters or colons can start the tag name.

If we merge the two, it will alter this rule.